### ISSUE TITLE
Question on Interoperatability of netopeer2-server and sysrepo API

### ISSUE THREAD

**Issue by alexyin1** [sysrepo/sysrepo#3077]
I'd like to create a network management system with the below flow
```
netconf_client <---> netopeer2_server <---> application
                 ^                      ^
            NETCONF RPC            sysrepo API
```
where netconf_client and netopeer2_server interact in the typical way (NETCONF rpc's), but the application interacts with sysrepo via sr_connect, sr_session_start, and sr_get_subtree to access configuration data.

In the above setup, is sysrepo sharable and thread-safe? Ex: is it possible to "share" the sysrepo connection and datastore between netopeer2_server and my application, such that if netopeer2_server receives an "edit-config" from the client, that the application can call "sr_get_subtree" and get data with the client's updated config?

I've looked around the [sysrepo plugin docs](https://netopeer.liberouter.org/doc/sysrepo/master/html/group__dsplg__api.html), but I not sure exactly how netopeer2_server interacts with sysrepo, and if netopeer2_server's sysrepo datastore is accessible from other processes.

Is my understanding of how to interface my application with netopeer2_server and sysrepo correct?

Thanks,
Alex

**Comment by jktjkt** [comment]
Yes, this is exactly how sysrepo is used in the real world. This picture from the homepage hopefully make it clear:

![image](https://github.com/sysrepo/sysrepo/assets/2631925/155625d5-7e12-4b35-8d9e-df6afb0f171c)

Your question mentions a "sysrepo connection" and you ask about its thread-safety and shareability. In the real world, Netopeer2 lives in one process, and each of your applications live in an extra process(es) of its own. One of the purposes of sysrepo is that it "somehow" manages this interaction, in this version through shared memory. Regarding thread safety, this is all [explained in the docs](https://netopeer.liberouter.org/doc/sysrepo/devel/html/index.html#threads).

**Comment by alexyin1** [comment]
Thanks for the detailed response! 

If I understand correctly, the IPC between application and netopeer2 is safe, because the shared memory files under [SR_SHM_DIR](https://github.com/sysrepo/sysrepo/blob/54c1b01537ca9909f18de202e0682a5be06f836b/src/config.h.in#LL80C9-L80C19) (by default /dev/shm I think) are managed by sysrepo.

Out of curiosity, how do the sysrepo API calls manage to safely share those SHM files? 
Suppose my app calls `sr_get_subtree(sr_sesh, xpath.c_str(), 0, &sr_data);` at the same time as when netopeer2 is processing an `edit-config` rpc. Is there a locking mechanism on SHM files? 

More specifically, are the functions in the [get_data API](https://netopeer.liberouter.org/doc/sysrepo/master/html/group__get__data__api.html#ga88900c0e398e1dba45c0b082f944b732) safe to use with netopeer2, or better to only use [subscribe API](https://netopeer.liberouter.org/doc/sysrepo/devel/html/subs.html) calls?

**Comment by michalvasko** [comment]
> Out of curiosity, how do the sysrepo API calls manage to safely share those SHM files?

By using our own synchronization primitives in those SHM segments that are based on mutexes and conditional variables/futexes.

> More specifically, are the functions in the [get_data API](https://netopeer.liberouter.org/doc/sysrepo/master/html/group__get__data__api.html#ga88900c0e398e1dba45c0b082f944b732) safe to use with netopeer2, or better to only use [subscribe API](https://netopeer.liberouter.org/doc/sysrepo/devel/html/subs.html) calls?

Ideally, any functions can be called concurrently and they should be synchronized to prevent any problems as needed. If it is not so, it is a bug that should be reported.

**Comment by alexyin1** [comment]
I see, thanks! Will give it a try.
